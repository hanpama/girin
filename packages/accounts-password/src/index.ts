// import { LocalAuth, LocalAuthConfigs } from './module';
// import { User } from './User';

// export default LocalAuth;
// export * from './decorators';
// export * from './User';
// export { LocalAuthConfigs };

