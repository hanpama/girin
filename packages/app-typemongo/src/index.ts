export * from './TypeMongoFrameworkDatastore';
export * from './createApp';
