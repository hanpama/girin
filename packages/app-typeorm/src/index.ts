export * from './connection';
export * from './createApp';
export * from './TypeORMFrameworkDatastore';
