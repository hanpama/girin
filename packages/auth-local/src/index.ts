import { LocalAuth, LocalAuthConfigs } from './module';
import { User } from './User';

export default LocalAuth;
export * from './decorators';
export * from './User';
export { LocalAuthConfigs };


export interface AuthContext<TUser extends User = User> {
  user: TUser;
}