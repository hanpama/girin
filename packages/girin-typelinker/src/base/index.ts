export * from './Definition';
export * from './MetadataStorage';
export * from './scalar';
export * from './structure';
export * from './TypeExpression';
