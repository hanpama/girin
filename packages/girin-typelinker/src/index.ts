export * from './base';
export * from './metadata';
export * from './field';
export * from './global';
