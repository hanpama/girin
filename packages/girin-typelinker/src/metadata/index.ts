export * from './InputType';
export * from './ObjectType';
export * from './InterfaceType';
export * from './rootTypes';
