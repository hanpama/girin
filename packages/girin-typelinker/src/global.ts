import { TypeArg, TypeExpression, MetadataStorage, loadBuiltInScalar, TypeExpressionKind, TypeExpressionConstructorOptions, Definition } from "./base";
import { SubstitutionMap, DefinitionParser } from "./sdl/ast";
import { parse } from "graphql";
import { ObjectType, InputType, InterfaceType, loadFallbackRootTypes } from "./metadata";


export function createMetadataStorage() {

  const storage = new MetadataStorage();
  loadBuiltInScalar(storage);
  loadFallbackRootTypes(storage);
  return storage;
}

/**
 * Global MetadataStorage used by default.
 */
export let globalMetadataStorage: MetadataStorage;

export function getGlobalMetadataStorage() {
  if (!globalMetadataStorage) {
    globalMetadataStorage = createMetadataStorage();
  }
  return globalMetadataStorage;
}

/**
 * Get a GraphQLType instance from the given storage or default
 * global metadata storage.
 * @param typeArg
 * @param storage
 */
export function getGraphQLType(typeArg: TypeArg, as: TypeExpressionKind = 'any', maybeStorage?: MetadataStorage): any {
  const storage = maybeStorage || getGlobalMetadataStorage();
  const typeExpression = new TypeExpression(typeArg, as);
  return typeExpression.getTypeInstance(storage!);
}

const SUBSTITUTION_PREFIX = '__GIRIN__SUBS__';

export function gql(strings: TemplateStringsArray, ...interpolated: Array<TypeExpression | TypeExpressionConstructorOptions>): DefinitionParser[] {
  const result = [strings[0]];
  const subsMap: SubstitutionMap = {};

  for (let i = 0; i < interpolated.length; i++) {
    const item = interpolated[i];
    let name: string;
    if (typeof item === 'string') {
      name = item;
    } else {
      name = `${SUBSTITUTION_PREFIX}${i}`;
      subsMap[name] = item;
    }
    result.push(name);
    result.push(strings[i + 1]);
  }

  const ast = parse(result.join(''));
  return ast.definitions.map(rootNode => new DefinitionParser(rootNode, subsMap));
}

// export interface TypeDefOptions {
//   defs: DefinitionParser[],
//   storage: MetadataStorage,
//   derive:
// }

// export function typedef(options: TypeDefOptions): void;
export function typedef(astParsers: DefinitionParser[], maybeStorage?: MetadataStorage) {
  const storage = maybeStorage || getGlobalMetadataStorage();

  return function defDecoratorFn(linkedClass: Function): void {
    astParsers.forEach(astParser => {

      astParser.objectTypeMetadataConfigs.forEach(config => {
        storage.register(new ObjectType(config), linkedClass);
      });
      astParser.inputObjectTypeMetadataConfigs.forEach(config => {
        storage.register(new InputType(config), linkedClass);
      });
      astParser.interfaceTypeMetadataConfigs.forEach(config => {
        storage.register(new InterfaceType(config), linkedClass);
      });
      astParser.implementTypeExpressions.forEach(exp => {
        if (astParser.extendingTypeName) {
          // later
        } else {
          storage.registerImplementReference({ definitionClass: linkedClass, interfaceType: exp });
        }
      });

      astParser.fieldMetadataConfigs.forEach(field => {
        const maybeStaticResolver = (linkedClass as any)[field.defaultName];
        const resolver = maybeStaticResolver instanceof Function ? maybeStaticResolver.bind(linkedClass) : undefined;
        if (astParser.extendingTypeName) {
          storage.registerExtensionFieldReference({ field, resolver, extendingTypeName: astParser.extendingTypeName });
        } else {
          storage.registerFieldReference({ field, definitionClass: linkedClass, resolver });
        }
      });
      astParser.inputFieldMetadataConfigs.forEach(config => {
        if (astParser.extendingTypeName) {
          storage.register
        } else {
          storage.registerInputFieldReference({ definitionClass: linkedClass, field: config })
        }
      });
    });
  }
}
