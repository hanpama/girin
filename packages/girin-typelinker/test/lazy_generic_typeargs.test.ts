import { GraphQLSchema, printSchema } from "graphql";

import { ObjectType, Definition, gql, getGraphQLType, List } from "../src";


@Definition.define(gql`
  type Edge {
    node: ${Edge => Edge.nodeType}
  }
`)
abstract class Edge<TNode> {
  static nodeType: Function;
  node: TNode;
}

@ObjectType.define(gql`
  type Post {
    title: String!
  }
`)
class Post {
  title: string;
}

@ObjectType.define(gql`
  type Author {
    name: String!
  }
`)
class Author {
  name: string;
}

@ObjectType.define()
class PostEdge extends Edge<Post> {
  static nodeType = Post;
}

@ObjectType.define()
class AuthorEdge extends Edge<Author> {
  static nodeType = Author;
}

@ObjectType.define(gql`
  type Query {
    postEdge: ${PostEdge}
    authorEdge: ${AuthorEdge}
    postEdges: ${List.of(PostEdge)}
    authorEdges: ${List.of(AuthorEdge)}
  }
`)
class Query {

}

describe('Lazy-generic TypeArgs', () => {
  it('should generate schema as expected', () => {
    const schema = new GraphQLSchema({
      query: getGraphQLType(Query),
    });

    expect(printSchema(schema)).toMatchSnapshot();

  });
});
