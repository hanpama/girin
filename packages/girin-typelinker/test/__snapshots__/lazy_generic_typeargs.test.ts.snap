// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy-generic TypeArgs should generate schema as expected 1`] = `
"type Author {
  name: String!
}

type AuthorEdge {
  node: Author
}

type Post {
  title: String!
}

type PostEdge {
  node: Post
}

type Query {
  postEdge: PostEdge
  authorEdge: AuthorEdge
  postEdges: [PostEdge]
  authorEdges: [AuthorEdge]
}
"
`;
